File: aoc-clj-2021/src/aoc_clj_2021/day6.clj
(ns aoc-clj-2021.day6
  (:require [clojure.string :as str]
            [clojure.set :as set]
            [clojure.java.io :as io]))

(def data (->> (io/resource "day6.txt") slurp (str/trim-newline)))
(def nums (map parse-long (str/split data #",")))

;; part 1
(-> (iterate (fn [nums] (mapcat #(if (zero? %) [6 8] [(dec %)]) nums))
             nums)
    (nth 80)
    count)

;; part 2
(def fish (merge (apply hash-map (interleave (range 9) (repeat 0)))
                 (frequencies nums)))

(defn evolve-fish [fish]
  (let [newfish (fish 0 0)]
    (->> fish
         (mapcat (fn [[i cnt]]
                   (cond (zero? i) [8 newfish]
                         (= 6 i) [6 (+ newfish (fish 7 0)) 5 (fish 6 0)]
                         :else [(dec i) (fish i 0)])))
         (apply hash-map))))


(reduce + (-> (iterate evolve-fish fish) (nth 256) (vals)))

File: aoc-py-2021/day6.py
data = open('day6.txt').read().split(',')
nums = [int(x) for x in data]

for i in range(80):
    new_nums = []
    for n in nums:
        if n == 0:
            new_nums.append(6)
            new_nums.append(8)
        else:
            new_nums.append(n-1)
    nums = new_nums

print(len(nums))

## part 2
nums = [int(x) for x in data]
ages = dict(zip(range(9),[0]*9))
for n in nums:
    ages[n] += 1

for i in range(256):
    newfish = ages[0]
    for age in range(1,9):
        ages[age-1] = ages[age]
    ages[6] += newfish
    ages[8] = newfish

print(sum(ages.values()))